src/builtins.c

```diff
@@ -1057,17 +1053,10 @@ static uintptr_t jl_object_id_(jl_value_t *tv, jl_value_t *v)
     jl_datatype_t *dt = (jl_datatype_t*)tv;
     if (dt == jl_datatype_type) {
         jl_datatype_t *dtv = (jl_datatype_t*)v;
-        uintptr_t h = 0xda1ada1a;
-        // has_typevars always returns 0 on name->primary, so that type
-        // can exist in the cache. however, interpreter.c mutates its
-        // typevars' `bound` fields to 0, corrupting the cache. this is
-        // avoided simply by hashing name->primary specially here.
-        if (jl_egal(dtv->name->primary, v))
-            return bitmix(bitmix(h, dtv->name->uid), 0xaa5566aa);
-        return bitmix(bitmix(h, dtv->name->uid), hash_svec(dtv->parameters));
+        return bitmix(~dtv->name->hash, hash_svec(dtv->parameters));
```

Comment:
might it be worthwhile to have a special case for emptysvec (microoptimization) for `jl_f_is` & hash?

```
return dta->name == dtb->name &&
    ((dta->parameters == jl_emptysvec) == (dtb->parameters == jl_emptysvec)) &&
    compare_svec(dta->parameters, dtb->parameters);
```